torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -180,6 +181,43 @@ def _post_forward(
     return output
 
 
+@no_type_check
+def _fsdp_root_pre_forward(
+    state: _State,
+    *args,
+    **kwargs,
+):
+    """
+    Runs pre-forward logic specific to the root FSDP instance, which should run
+    before any individual module's pre-forward. If this is called on a non-root
+    FSDP instance, then the forward inputs are returned directly.
+    """
+    p_assert(state._is_root is not None, "Expects a root FSDP to have been set")
+    if not state._is_root:
+        return args, kwargs
+    if state.forward_prefetch:
+        handles_keys = []
+        if _is_composable(state):
+            # TODO: This assumes singleton handles keys.
+            handles_keys = [tuple(handle) for handle in state._handles]
+        else:
+            for fsdp_module in state.fsdp_modules(state):
+                handles_key = tuple(fsdp_module._handles)
+        for handles_key in handles_keys:
+            state._needs_pre_forward_unshard[handles_key] = True
+    _wait_for_computation_stream(
+        torch.cuda.current_stream(),
+        state._streams["unshard"],
+        state._streams["pre_unshard"],
+    )
+    _clear_grads_if_needed(_all_handles(state))
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
```

Comment:
hmm, this part looks a bit different, it used to be:

```python
        input_dtype = (
            self.mixed_precision.param_dtype
            if self._mixed_precision_enabled_for_params()
            else None
        )
```

Reply:
Yup. This is intended and just a simplification. `self._mixed_precision_enabled_for_params()` is exactly just `return self.mixed_precision.param_dtype is not None`.